# Analyzing Used Car Listings on eBay Kleinanzeigen

I will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The dataset was originally scraped and uploaded to Kaggle. 
The data dictionary provided with data is as follows:
- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [35]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/used-cars-database/autos.csv
/kaggle/input/used-cars-database/cnt_km_year_powerPS_minPrice_maxPrice_avgPrice_sdPrice.csv


In [36]:
autos = pd.read_csv('/kaggle/input/used-cars-database/autos.csv', sep=',', header=0, encoding='cp1252')
#df = pd.read_csv('autos.csv.gz', sep=',', header=0, compression='gzip',encoding='cp1252')
autos.sample(10)

dateCrawled  \
16930   2016-03-22 17:44:42   
101205  2016-03-28 13:46:00   
115187  2016-03-15 08:55:46   
209293  2016-03-15 17:53:30   
285527  2016-04-04 12:06:22   
335276  2016-03-14 20:38:22   
16301   2016-04-04 23:50:45   
323881  2016-03-18 20:22:04   
100019  2016-03-28 19:39:50   
186550  2016-03-17 16:49:05   

                                                     name  seller offerType  \
16930               Audi_A6_4b_3.0_V6_mit_Getriebeproblem  privat   Angebot   
101205                      Volkswagen_Golf_1.4_Trendline  privat   Angebot   
115187                                           BMW_320i  privat   Angebot   
209293                                 Ford_Focus_TÜV_neu  privat   Angebot   
285527                                    Bastlerfahrzeug  privat   Angebot   
335276  A4_Avant_2.0_TDI__S_line_Sportpaket_plus_Navi_...  privat   Angebot   
16301   Smart_ForTwo_CDI_Cabrio__viele_EXTRAS_nur_3_5_...  privat   Angebot   
323881  2.Hand_Schiebedach_30_Jahre_Sonderedition_Klim...  privat   Angebot   
100019  Mercedes_Benz_C_250_AMG_Sport_CDI_BE_Coupe_Aut...  privat   Angebot   
186550  Audi_A3_2.0_TDI_Sportback_DPF_S_line_Sportpake...  privat   Angebot   

        price   abtest vehicleType  yearOfRegistration    gearbox  powerPS  \
16930    3000  control       kombi                2002  automatik      220   
101205   6750     test   limousine                2009    manuell       80   
115187   1850     test   limousine                1991    manuell      150   
209293    899  control       kombi                1999    manuell      125   
285527    250  control      andere                1989    manuell        0   
335276   4200  control       kombi                2007    manuell      170   
16301    9400  control      cabrio                2013  automatik       54   
323881   4890     test   limousine                2007    manuell       95   
100019  24500  control       coupe                2011  automatik      204   
186550   4300     test       kombi                2005    manuell      140   

           model  kilometer  monthOfRegistration fuelType          brand  \
16930         a6     125000                    2   benzin           audi   
101205      golf     125000                    6   benzin     volkswagen   
115187       3er     150000                    3   benzin            bmw   
209293     focus     150000                   10   benzin           ford   
285527      golf     150000                    0   benzin     volkswagen   
335276        a4     150000                    4   diesel           audi   
16301     fortwo      70000                    5   diesel          smart   
323881      colt     100000                    7   benzin     mitsubishi   
100019  c_klasse      90000                    7   diesel  mercedes_benz   
186550        a3     150000                    7   diesel           audi   

       notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
16930                 ja  2016-03-22 00:00:00             0       76448   
101205              nein  2016-03-28 00:00:00             0       84559   
115187              nein  2016-03-15 00:00:00             0       35066   
209293                ja  2016-03-15 00:00:00             0       46045   
285527                ja  2016-04-04 00:00:00             0       35644   
335276              nein  2016-03-14 00:00:00             0        6886   
16301               nein  2016-04-04 00:00:00             0       41836   
323881               NaN  2016-03-18 00:00:00             0       65197   
100019              nein  2016-03-28 00:00:00             0       52064   
186550                ja  2016-03-17 00:00:00             0       51588   

                   lastSeen  
16930   2016-03-24 01:44:48  
101205  2016-03-31 21:46:47  
115187  2016-03-16 04:16:50  
209293  2016-03-16 15:44:41  
285527  2016-04-05 12:16:18  
335276  2016-03-14 20:38:22  
16301   2016-04-07 08:46:53  
323881  2016-04-02 13:17:55  
100019  2016

In [37]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

we can make the following observations:
- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [38]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:
- Change the columns from camelcase to snakecase.
- Change a few wordings to more accurately describe the columns.

In [39]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

# Initial Data Exploration and Cleaning
We'll start by exploring the data to find obvious areas where we can clean the data.

In [40]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

My initial observations:
There are a number of text columns where all (or nearly all) of the values are the same:
- seller
- offer_type

The num_photos column looks odd, I'll need to investigate this further.

In [41]:
autos["num_photos"].value_counts()

0    371528
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. I'll drop this column, plus the other two we noted as mostly one value.

In [42]:
autos=autos.drop(['num_photos','seller','offer_type'],axis=1)

# Exploring Odometer and Price

In [43]:
autos.rename({'odometer':'odometer_km'},axis=1, inplace=True)

In [44]:
autos["odometer_km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [45]:
print('There are',autos['price'].unique().shape[0],'unique values in the price column.')
print('\n')
print(autos['price'].describe())
print('\n')
print(autos['price'].value_counts().head(10))

There are 5597 unique values in the price column.


count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
Name: price, dtype: int64


Again, the prices in this column seem rounded, however given there are 5597 unique values in the column, that may just be people's tendency to round prices on the site.

There are 10,778 cars listed with $0 price - given that this is less than 3% of the of the cars, we might consider removing these rows. 

The maximum price is 2 billion dollars, which seems a lot, let's look at the highest prices further.

In [46]:
autos["price"].value_counts().sort_index(ascending=False).head(30)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
1111111        2
1010010        1
1000000        5
999999        13
999990         1
911911         1
849000         1
820000         1
780000         1
745000         2
Name: price, dtype: int64

In [47]:
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 84 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [48]:
autos=autos[autos['price'].between(1,351000)]
autos["price"].describe()

count    360635.000000
mean       5898.671956
std        8866.359669
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the date columns
There are a number of columns with date information:
- date_crawled
- registration_month
- registration_year
- ad_created
- last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [49]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [50]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001617
2016-04-06    0.003128
2016-04-05    0.012780
2016-03-18    0.013119
2016-03-06    0.014483
2016-03-13    0.015783
2016-03-05    0.025547
2016-03-24    0.029914
2016-03-16    0.030205
2016-03-27    0.030227
2016-03-17    0.031647
2016-03-31    0.031872
2016-03-26    0.031974
2016-03-23    0.032002
2016-03-22    0.032493
2016-03-10    0.032645
2016-03-11    0.032773
2016-03-25    0.032800
2016-03-15    0.033424
2016-03-08    0.033469
2016-03-30    0.033535
2016-03-09    0.034115
2016-03-29    0.034126
2016-04-01    0.034145
2016-03-28    0.035063
2016-04-02    0.035094
2016-03-19    0.035271
2016-03-07    0.035657
2016-03-21    0.035682
2016-03-12    0.036242
2016-03-14    0.036330
2016-03-20    0.036400
2016-04-04    0.037628
2016-04-03    0.038812
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [51]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023240
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028699
2016-03-18    0.006888
2016-03-19    0.016330
2016-03-20    0.019884
2016-03-21    0.020026
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016721
2016-03-28    0.022189
2016-03-29    0.023284
2016-03-30    0.023725
2016-03-31    0.024243
2016-04-01    0.023897
2016-04-02    0.024967
2016-04-03    0.025308
2016-04-04    0.025536
2016-04-05    0.126962
2016-04-06    0.218950
2016-04-07    0.130437
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.
The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [52]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039001
2016-04-04    0.037736
2016-04-05    0.011613
2016-04-06    0.003119
2016-04-07    0.001553
Name: ad_created, Length: 114, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [53]:
autos["registration_year"].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

# Dealing with Incorrect Registration Year Data

Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [54]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038751091824143526

Given that this is less than 4% of our data, we will remove these rows.

In [55]:
# Many ways to select rows in a dataframe that fall within a value range for a column.
# Using `Series.between()` is one way.
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.066699
1999    0.063552
2005    0.062669
2006    0.057708
2001    0.056955
2003    0.056557
2004    0.056173
2002    0.054290
2007    0.050499
1998    0.049691
Name: registration_year, dtype: float64

# Translate German values into English

In [56]:
autos['vehicle_type'].value_counts()

limousine     93826
kleinwagen    78200
kombi         66085
bus           29764
cabrio        22550
coupe         18447
suv           14499
andere         3150
Name: vehicle_type, dtype: int64

In [57]:
di={'kleinwagen':'small car','andere':'other'}
autos = autos.replace({"vehicle_type": di})

In [58]:
autos['gearbox'].value_counts()

manuell      257012
automatik     73725
Name: gearbox, dtype: int64

In [59]:
di={'manuell':'manual','automatik':'automatic'}
autos = autos.replace({"gearbox": di})

In [60]:
autos['fuel_type'].value_counts()

benzin     212127
diesel     103815
lpg          5090
cng           539
hybrid        265
andere        167
elektro       101
Name: fuel_type, dtype: int64

In [61]:
di={'elektro':'electro','andere':'other','benzin':'petrol'}
autos=autos.replace({'fuel_type':di})

# Exploring Price by Brand

In [62]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211700
bmw               0.109871
opel              0.106410
mercedes_benz     0.096841
audi              0.089543
ford              0.068918
renault           0.047516
peugeot           0.030153
fiat              0.025691
seat              0.018661
skoda             0.015687
mazda             0.015384
smart             0.014331
citroen           0.013950
nissan            0.013598
toyota            0.012932
hyundai           0.009972
sonstige_autos    0.009493
mini              0.009384
volvo             0.009147
mitsubishi        0.008236
honda             0.007532
kia               0.006915
suzuki            0.006364
alfa_romeo        0.006309
porsche           0.006211
chevrolet         0.005022
chrysler          0.003863
dacia             0.002495
jeep              0.002192
land_rover        0.002166
daihatsu          0.002161
subaru            0.002117
jaguar            0.001734
saab              0.001465
daewoo            0.001457
trabant           0.001408
l

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [63]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [64]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5400,
 'bmw': 8449,
 'opel': 2971,
 'mercedes_benz': 8551,
 'audi': 9086,
 'ford': 3696}

# Exploring Mileage

In [65]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5400
bmw                  8449
opel                 2971
mercedes_benz        8551
audi                 9086
ford                 3696

In [66]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [67]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132800
mercedes_benz        130572
audi                 129443
opel                 128722
volkswagen           128386
ford                 123662

In [68]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132800        8449
mercedes_benz        130572        8551
audi                 129443        9086
opel                 128722        2971
volkswagen           128386        5400
ford                 123662        3696

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.